### PR TITLE
improve stretch histogram performance (downsampling and decorator)

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request improves the performance of the stretch histogram and the stretch curve, especially for large images, by downsampling the data sent to the histogram and by making use of #2386 for the stretch curve introduced in #2390.

https://github.com/spacetelescope/jdaviz/assets/877591/3a99859d-2b85-4e33-a8a1-517aa837bbb6

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 1]
```diff
@@ -600,7 +594,18 @@ def _update_stretch_histogram(self, msg={}):
...
                 y_max = y_limits.max()
 
-                sub_data = comp.data[y_min:y_max, x_min:x_max].ravel()
+                arr = comp.data[y_min:y_max, x_min:x_max]
+
+                size = arr.shape[0] * arr.shape[1]
+                if size > 400**2:
+                    xstep = max(1, round(arr.shape[1] / 400))
+                    ystep = max(1, round(arr.shape[0] / 400))
+                    arr = arr[::ystep, ::xstep]
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:603]: @pllim - this code block essentially exists in three places in jdaviz now (compass and twice here).  Do you think it's worth moving somewhere reusable?  Maybe in utils?

**Comment by camipacifici** [comment]
Great improvement, thank you!
I wonder whether the sampling could be avoided under a certain number of pixels, like when zooming into a specific region of a big image.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:603]: Not sure. If you want this to turn into utility function, then you have to make sure that function is not making accidental copies (everything must pass by reference where it matters) and maybe a way to adjust sampling factor and maybe even indices (for generic ndim support). Something to think about.

Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:603]: Do you want to refactor in this PR? If so, I will wait before reviewing.

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:603]: I had some of those same concerns, so let's leave this as-is for now and consider moving the logic up higher if we need it anywhere else in the future.

Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:603]: I have a bad feeling that we might have to revisit this as more and more new analysis stuff get requested. Should we open a ticket pre-emptively for the future?

**Review by pllim — COMMENTED** [review]
I think it is unavoidable to decrease the accuracy of histogram to make it not lag so much. I tried comparing like this and while I see some difference, it is good enough. This is an example using the first image in Imviz Example Notebook when zoomed in and only show data within display window.

Screenshot 2023-08-30 171732

```python
# Approx X/Y limits from imviz.default_viewer.state.x/y_min/max
pf = fits.open(f'{data_dir}/jw02727-o002_t062_nircam_clear-f090w_i2d.fits')
plt.hist(pf[1].data[2131:3523, 1699:3244].ravel(),
         bins=25, range=(0, 0.5852748259901993),
         rwidth=0.9, color="gray")
plt.axvline(0, color="red")
plt.axvline(0.5852748259901993, color="red")
```

Screenshot 2023-08-30 172236

But can we make the blue info box a little smaller?

Screenshot 2023-08-30 170349

**Comment by pllim** [comment]
So does histogram even need Y tick marks at this point? The numbers on Y axis do not mean much here.

**Comment by kecnry** [comment]
> So does histogram even need Y tick marks at this point? The numbers on Y axis do not mean much here.

They are normalized density, so I wouldn't say they're meaningless, but I'd also be in favor of removing them since I think that would be assumed (or we can keep the label, but remove the numbers).

> But can we make the blue info box a little smaller?

I don't want to set specific rules for this alert that differs from the rest.... but let me try alternate options for how to display this information.  Maybe even in the title of the plot (so that it also shows in the plot popout...)?

**Comment by pllim** [comment]
> They are normalized density, so I wouldn't say they're meaningless

Not much meaning when it is subsampling. I guess it is fine if you want to keep the numbers but I fear they might confuse more than help. Unless you want to normalize to max of 1. Then it can be like percentage but in decimal form.

**Review by pllim — APPROVED** [review]
Cami is happy. Code looks like code.

Y-axis discussions is technically out of scope, so approving. Thanks!

**Pull request merged**